### ISSUE TITLE
QUESTION: usecase of test junkie on flask app without UI

### ISSUE THREAD

**Issue by aditiabhang** [ArturSpirin/test_junkie#24]
I am a computer science graduate student and I have chosen to study and make a report on your testing framework 'test-junkie' as part of my coursework on software quality. So far I have started to go through the documentation and tutorials mentioned on the website. However, what I observed from the tutorial examples is that you are using the framework to test components of your website. I am fairly new to python and I have no experience in testing UI components of a website.

I am curious to understand on how I can use test-junkie on a flask based app in python. I am little confused if test-junkie library needs to be imported directly in the main application code or do I import it inside my unit tests or we do not write unit tests at all and use test junkie as a substitute for it?

I would really appreciate your insights to help me understand the use-case of test-junkie as it would really help me with my project.

My apologies for opening up a conversation in a github issue rather than email but I did not find your email address anywhere on the website.

**Comment by ArturSpirin** [comment]
Hi,

Thanks for your interest in Test Junkie! Idealy this question should have been open in a StackOverflow but I'm not that active on there and don't have enough reputation to create a new topic for it so this is OK!

I'm going to use Unittest standard Python library to try to answer your questions.

Now, lets just calarify a few terms:
- End-to-end testing is typically a test of user flow this test may or may not be part of the same code base where you have your application. Ex: login to a platform, adding an item to shoping cart, or performing a search etc
- Unit testing is typically a test of a function or a class within the code base of the applciation e: a function that makes a call to a database to get user data

The use cases of Test Junkie are exactly the same as use cases of the standard Unittest library in Python. It can be used as a unit testing framework or an end-to-end testing framework. The major difference between the two is the configuration options and the features which make it easier to configure and scale tests. At its core niether of the frameworks really care what you are testing, weather its API via HTTP requests, a function call in Python code, a UI operation in a browser or anything else. Both of the frameworks are generic that they can be used to test anything. Now, ofcorse if you have a PHP web application you cannot use Test Junkie to write unit tests for that PHP application, a PHP testing framework would be prefered here. However, if you have a Flask application, you can use Test Junkie for unit testing. In both cases of PHP and Flask web application, you can use Test Junkie to test the client side of the application this is more or less the end-to-end user flows that are performed from the browser, this is exactly what I do in my tests that are features on the https://www.test-junkie.com/tutorials/.

Typically when organizations start investing into test automation, the test engineers start writing very basic tests and test architecture to capture the low hanging fruits as quickly as possible. Once that is done, they usually end up with a regid architecture that have many limits and requires much of refactoring to lift them. Test Junkie tries to solve this issues by providing all configuration options in the decorators. So if there is anything you want to change, all it takes is a minor tweak of the docorator properties.

For instance, if you are using Unittest framework, to run tests in parallel you need to build everything from the ground up but in Test Junkie, its not evern an extra line of code - its just a flag.

If there is anything else I can try to help and answer, please let me know!

**Comment by aditiabhang** [comment]
Thanks for clarifying on the use case question. Since scope of my work is using test-junkie to write unit tests and study it in terms of the ease of use. However, I do not have a big project where I can test different components of it for end-to-end test. Although, for now this helps me understand on how I can write unit tests for my small codes. I appreciate for your prompt reply. I will post a question in case I have any further questions : )

**Comment by ArturSpirin** [comment]
Always happy to help!

**Comment by aditiabhang** [comment]
Good Morning Arthur,
I have been trying to understand the code coverage feature of Test Junkie,  by executing a simple python program. I got resulting output with multiple lines of details which I am finding it quite confusing. Could you please help me understand the code coverage result? 

I would be very thankful for your guidance. 

I am attaching the snapshot of my resulting code coverage, please find the attachment. 
<img width="1202" alt="Screen Shot 2019-11-20 at 11 09 45 AM" src="https://user-images.githubusercontent.com/51350594/69260926-489d1a00-0b86-11ea-8c39-7ecd120d8961.png">

**Comment by ArturSpirin** [comment]
Hello @aditiabhang 
FIrst I'm gonna ask you to update Test Junkie to v0.7a8, it has an important fix (#22 ) for coverage.

Typically coverage will be used to measure how much of the code base is covered by unit tests. So lets assume we have the following project:

```
demo_project
|_src
  |_Utility.py
|_tests
  |_TestSuite.py
```
Lets say this is the code in our Utility.py that does something amazing
```python


class Utility:

    def __init__(self):

        pass

    def do_something_1(self, value):
        print("Class: {} Function: {} Value: {}".format(self.__class__, self, value))

    def do_something_2(self, value):
        print("Class: {} Function: {} Value: {}".format(self.__class__, self, value))

    def do_something_3(self, value):
        print("Class: {} Function: {} Value: {}".format(self.__class__, self, value))

    def do_something_4(self, value):
        print("Class: {} Function: {} Value: {}".format(self.__class__, self, value))

```

Now we want to unit test it and see how much of the code we have covered with our tests, for this we write tests which look like the following

```python
from test_junkie.decorators import Suite, test

from src.Utility import Utility


@Suite()
class TestSuite:

    @test()
    def some_test_1(self):

        Utility().do_something_1(self)

    @test()
    def some_test_2(self):

        Utility().do_something_2(self)

    @test()
    def some_test_3(self):

        Utility().do_something_3(self)
```

Note that we did not test the method in the Utility class for `do_something_4` so we should be missing some coverage, lets run the tests using the CLI and see what we get.
CLI Command: ` tj run -s E:\Development\demo_project\tests\TestSuite.py --code-cov`
And the output
```bash
========== Test Junkie finished in 1.00 seconds ==========
[3/3 100.00%] SUCCESS
[0/3 0%] FAIL
[0/3 0%] IGNORE
[0/3 0%] ERROR
[0/3 0%] SKIP
[0/3 0%] CANCEL

>> [SUCCESS] [3/3 100.00%] [0.00s] TestSuite.TestSuite

===========================================================
. Test Junkie 0.7a8 (Python2) https://www.test-junkie.com .
===========================================================
[INFO] Code coverage report:
Name                                          Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------
E:\Development\demo_project\src\Utility.py      11      1    91%   19
---------------------------------------------------------------------------
TOTAL                                            20      1    95%

2 files skipped due to complete coverage.
[TIP] TJ uses Coverage.py. Control it with --cov-rcfile, see https://coverage.readthedocs.io/en/v4.5.x/config.html
```
So the output tells me that my unit tests do not cover all my code in the Utility class and now I will have to check what is not covered. If you look at the Missing column in the Coverage output, it tells you that code at line 19 was not executed and that to be expected. 

Hope this helps.
